libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1658 @@
+        source: source.chars().peekable(),
+        tokens: Vec::new(),
+        current_string: String::new(),
+        };
+    lexer.lex_tokens(); 
+    lexer.tokens
+}
+
+// StateFn maintains the lexers recursive state
+struct StateFn{ f: fn(&mut Lexer) -> Option<StateFn> }
+
+impl Lexer<'_> {
+    // lex_tokens recursively calls the lexer while there are still characters to lex
+    fn lex_tokens(&mut self){
+        let mut state_fn = Some(StateFn{ f: lex_token }); // assigning initial state
+
+        while let Some(func) = state_fn {
+            state_fn = (func.f)(self)
+        }
+    }
+
+    // next grabs the next character from the string while there are still characters to lex
+    fn next(&mut self) -> Option<char> {
+        match self.source.next() {
+            None => None, 
+            Some(letter) => {
+                if letter.is_alphanumeric() {
+                    self.current_string.push(letter);
+                }
+                Some(letter)
+            }
+        }
+    }
+
+    // emit instatiates a Token and pushes that token along with its TokenType and text, if applicable,
+    // into the Lexer's tokens vector
+    fn emit(&mut self, token: TokenType) {
+        if !self.current_string.is_empty() {
+            let text = self.current_string.clone();
+            self.tokens.push(Token{ token_type: token, text: Some(text) })
+        } else {
+            self.tokens.push(Token{ token_type: token, text: None })
+        }
+        self.current_string = String::new();
+    }
+
+    // ignore empties self.current_string. This is used for whitespace tokens since they are not currently
+    // taken into consideration when parsing
+    fn ignore(&mut self) {
+        self.current_string = String::new();
+    }
+
+    fn keyword_or_ident(&mut self) {
+        let mut keyword_match = true; 
+
+        while let Some(letter) = self.source.peek() {
+            if letter.is_alphanumeric() {
+                self.next();
+                continue; 
+            }
+            break; // if letter is not alphanumeric
+        }
+
+        let current: &str = &self.current_string; 
+        match current {
+            "forall" => { self.emit(TokenType::FORALL); },
+            "where" => { self.emit(TokenType::WHERE); },
+            "int" => { self.emit(TokenType::INT); },
+            "float" => { self.emit(TokenType::FLOAT); },
+            "string" => { self.emit(TokenType::STRING); },
+            "bool" => { self.emit(TokenType::BOOL); },
+            "uint" => { self.emit(TokenType::UINT); },
+            "duration" => { self.emit(TokenType::DURATION); },
+            "time" => { self.emit(TokenType::TIME); },
+            "regexp" => { self.emit(TokenType::REGEXP); },
+            _ => { keyword_match = false; } 
+        }
+        
+        if !keyword_match {
+            // token is not a keyword, must be an identifier
+            self.emit(TokenType::IDENTIFIER);
+        }        
```

Comment:
Wow, this is great. Much simpler thanks! 

One last simplification now is that the `keyword_match` boolean is not needed as this last if expression can be moved inside the `_ => ` section of the match expression.